File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/HistoricElementSecurityError.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/HistoricElementSecurityError.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import com.bloomberglp.blpapi.Element;
import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.SchemaTypeDefinition;

public class HistoricElementSecurityError extends Element
{
    private final HistoricElementString _source, _category, _message, _subCategory;
    private final HistoricElementInt _code;

    HistoricElementSecurityError(String security)
    {
        int code = RandomDataGenerator.randomInt(99);
        String sourceGibberish = RandomDataGenerator.randomString(5).toLowerCase();

        this._source = new HistoricElementString("source", String.format("%s::%s%s", code, sourceGibberish, RandomDataGenerator.randomInt(99)));
        this._code = new HistoricElementInt("code", code);
        this._category = new HistoricElementString("category", "BAD_SEC");
        this._message = new HistoricElementString("message", String.format("Unknown/Invalid security [nid:%s]", code));
        this._subCategory = new HistoricElementString("subcategory", "INVALID_SECURITY");
    }
    
    public SchemaTypeDefinition typeDefinition() throws Exception
    {
    	return new SchemaTypeDefinition(this.datatype(), new Name("ErrorInfo"));
    }
	
    public Name name() throws Exception
    {
    	return new Name("securityError");
    }
    
    public int numValues()
    {
    	return 1;
    }
    
    public int numElements()
    {
    	return 5;
    }
    
    public String getElementAsString(String name) throws Exception
    {
    	return this.getElement(name).getValueAsString();
    }
    
    public int getElementAsInt32(String name) throws Exception
    {
    	return this.getElement(name).getValueAsInt32();
    }
    
    public Element getElement(String name) throws Exception
    {
    	if(name.equals("source"))
    		return this._source;
    	
    	else if(name.equals("code"))
    		return this._code;
    	
    	else if(name.equals("category"))
    		return this._category;
    	
    	else if(name.equals("message"))
    		return this._message;
    	
    	else if(name.equals("subcategory"))
    		return this._subCategory;
    	
    	else
    		return super.getElement(name);
    }
    
    public boolean hasElement(String name)
    {
    	return name.equals("source") ||
    			name.equals("code") ||
    			name.equals("category") ||
    			name.equals("message") ||
    			name.equals("subcategory");
    }  
    
    protected StringBuilder prettyPrint(int tabIndent) throws Exception
    {
        String tabs = com.bloomberglp.blpapi.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();
        
        result.append(String.format("%s%s = {%s", tabs, this.name().toString(), System.getProperty("line.separator")));
        result.append(this._source.prettyPrint(tabIndent + 1));
        result.append(this._code.prettyPrint(tabIndent + 1));
        result.append(this._category.prettyPrint(tabIndent + 1));
        result.append(this._message.prettyPrint(tabIndent + 1));
        result.append(this._subCategory.prettyPrint(tabIndent + 1));
        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));

        return result;
    }
    
    public boolean isComplexType()
    {
    	return true;
    }
    
    public boolean isArray()
    {
    	return false;
    }
    
    public boolean isNull()
    {
    	return false;
    }

}

File: BEmu_cpp/src/HistoricalDataRequest/HistoricElementSecurityError.cpp
//------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/HistoricalDataRequest/HistoricElementSecurityError.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

#include "HistoricalDataRequest/HistoricElementSecurityError.h"
#include "HistoricalDataRequest/HistoricElementString.h"
#include "HistoricalDataRequest/HistoricElementInt.h"

#include "Types/RandomDataGenerator.h"
#include "Types/IndentType.h"
#include "BloombergTypes/Name.h"

namespace BEmu
{
	namespace HistoricalDataRequest
	{
		HistoricElementSecurityError::HistoricElementSecurityError(const std::string& security) :
			_code(new HistoricElementInt("code", RandomDataGenerator::RandomInt(99))),
			_category(new HistoricElementString("category", "BAD_SEC")),
			_subCategory(new HistoricElementString("subcategory", "INVALID_SECURITY"))
		{
			int nid = RandomDataGenerator::RandomInt(99);

			std::string sourceGibberish = RandomDataGenerator::RandomString(5);
			std::transform(sourceGibberish.begin(), sourceGibberish.end(), sourceGibberish.begin(), ::tolower);

			//source
			std::stringstream sourceValue;
			sourceValue << nid << "::" << sourceGibberish.substr(0, sourceGibberish.length() - 2) << RandomDataGenerator::RandomInt(99);
			this->_source = boost::shared_ptr<HistoricElementString>(new HistoricElementString("source", sourceValue.str()));

			//message
			std::stringstream messageValue;
			messageValue << "Unknown/Invalid security [nid:" << nid << "]";
			this->_message = boost::shared_ptr<HistoricElementString>(new HistoricElementString("message", messageValue.str()));
		}

		HistoricElementSecurityError::~HistoricElementSecurityError()
		{
		}


		Name HistoricElementSecurityError::name() const
		{
			Name result("securityError");
			return result;
		}

		size_t HistoricElementSecurityError::numValues() const
		{
			return 1;
		}

		size_t HistoricElementSecurityError::numElements() const
		{
			return 5;
		}

		SchemaElementDefinition HistoricElementSecurityError::elementDefinition() const
		{
			::blpapi_DataType_t dtype = (::blpapi_DataType_t)this->datatype();
			SchemaElementDefinition result(dtype, Name("ErrorInfo"));
			return result;
		}

		
		bool HistoricElementSecurityError::isNull() const
		{
			return false;
		}

		bool HistoricElementSecurityError::isArray() const
		{
			return false;
		}

		bool HistoricElementSecurityError::isComplexType() const
		{
			return true;
		}


		boost::shared_ptr<ElementPtr> HistoricElementSecurityError::getElement(const char* name) const
		{
			if(strncmp(name, "source", 7) == 0)
				return boost::dynamic_pointer_cast<ElementPtr>(this->_source);

			else if(strncmp(name, "code", 5) == 0)
				return boost::dynamic_pointer_cast<ElementPtr>(this->_code);

			else if(strncmp(name, "category", 9) == 0)
				return boost::dynamic_pointer_cast<ElementPtr>(this->_category);

			else if(strncmp(name, "message", 8) == 0)
				return boost::dynamic_pointer_cast<ElementPtr>(this->_message);

			else if(strncmp(name, "subcategory", 12) == 0)
				return boost::dynamic_pointer_cast<ElementPtr>(this->_subCategory);

			else
				throw elementPtrEx;
		}

		bool HistoricElementSecurityError::hasElement(const char* name, bool excludeNullElements) const
		{
			return
				(strncmp(name, "source", 7) == 0) ||
				(strncmp(name, "code", 5) == 0) ||
				(strncmp(name, "category", 9) == 0) ||
				(strncmp(name, "message", 8) == 0) ||
				(strncmp(name, "subcategory", 12) == 0);
		}


		int HistoricElementSecurityError::getElementAsInt32(const char* name) const
		{
			return this->getElement(name)->getValueAsInt32(0);
		}

		const char* HistoricElementSecurityError::getElementAsString(const char* name) const
		{
			return this->getElement(name)->getValueAsString(0);
		}


		std::ostream& HistoricElementSecurityError::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			std::string tabs(IndentType::Indent(level, spacesPerLevel));

			stream << tabs << "fieldData = {" << std::endl;
			
			this->_source->print(stream, level, spacesPerLevel);
			this->_code->print(stream, level, spacesPerLevel);
			this->_category->print(stream, level, spacesPerLevel);
			this->_message->print(stream, level, spacesPerLevel);
			this->_subCategory->print(stream, level, spacesPerLevel);

			stream << tabs << '}' << std::endl;
			
			return stream;
		}

	}
}